scc/committee.go

```diff
@@ -0,0 +1,172 @@
+package scc
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/0xsoniclabs/sonic/scc/bls"
+)
+
+// Committee is a group of Members that can sign statements to produce
+// certificates.
+//
+// On the Sonic Certification Chain (SCC), time is divided into periods. For
+// each period, a committee is authorized to sign the statement for that period.
+// Among those statements is the confirmation of the succeeding period's
+// committee.
+//
+// Committees are composed of a ordered list of Members, each weighted with a
+// non-zero voting power. Members are identified by their public keys, for which
+// they are required to provide a proof of possession.
+type Committee struct {
+	members []Member
+}
+
+// MaxCommitteeSize is the maximum number of members that can be in a committee.
+// The number needs to be limited to prevent Committee certificates from becoming
+// too large.
+const MaxCommitteeSize = 512
+
+// MemberId is used to identify a member in a committee by its position in the
+// ordered list of members.
+type MemberId uint16
+
+// NewCommittee creates a new committee with the provided members. There must be
+// at least one member, all members must be valid, and no duplicates are allowed.
+func NewCommittee(members ...Member) (Committee, error) {
+	res := Committee{members: slices.Clone(members)}
+	if err := res.Validate(); err != nil {
+		return Committee{}, err
+	}
+	return res, nil
+}
+
+// Members returns a copy of the members in the committee.
+func (c Committee) Members() []Member {
+	return slices.Clone(c.members)
+}
+
+// GetMember returns the member with the given id. If the id is out of bounds,
+// the second return value is false.
+func (c Committee) GetMember(id MemberId) (Member, bool) {
+	if int(id) >= len(c.members) {
+		return Member{}, false
+	}
+	return c.members[id], true
+}
+
+// GetMemberId returns the id of the member with the given public key. If the
+// public key is not found, the second return value is false.
+func (c Committee) GetMemberId(publicKey bls.PublicKey) (MemberId, bool) {
+	for id, m := range c.members {
```

Comment:
nit: `slices.IndecFunc` ?

Reply:
This would not really be shorter or more readable:
```
       res := slices.IndexFunc(c.members, func(m Member) bool {
		return m.PublicKey == publicKey
	})
	if res < 0 {
		return 0, false
	}
	return MemberId(res), true
```